### PR TITLE
refactor(reports): update balance report

### LINKED ISSUE #2617: Improvements to the "Balance of Accounts" report
We have an older report called "Balance of Accounts" which gives the balance of all accounts (independent of account type).  The numbers in the report agree with the General Ledger, however the presentation can be improved in the following ways.  See the below screenshot for reference:

balanceofaccountsproblem
_Fig 1: The Balance of Accounts as presented by the System_

**Recommendations**
 1. Order the accounts by account number.
 2. Include title accounts in bold and indent children.
 3. Provide an option to show/filter zero balanced accounts.  Currently only accounts with a balance are show.
 4. By default display the balance in the enterprise currency, but provide an option to use a different currency with an exchange rate prominently displayed.
 5. Toggle the view into separate debits and credits columns or a single value column that shows the debit and credit with a coloration difference.
 6. Select a period instead of a date.
 7. Implement a "check row" in the footer of the report.  It should be the sum of all rows and always be 0.

### PR DESCRIPTION
**Pull request by jniles**
This commit updates the balance report with a tree structure, an optional expanded or collapsed presentation and using a fiscal period select component instead of a naked date.

Closes #2617.

This report is now ready for review.

The interface looks like this:

**Configuration Panel**
balancereportconfiguration

**View of Combined Debits and Credits**
combineddebitsandcredits

**View of Separated Debits and Credits**
separateddebitsandcredits

### CODE REVIEW

File: client/src/js/components/bhReportPeriodSelect.js [hunk 1]
```diff
@@ -49,7 +49,5 @@ function ReportPeriodSelectController(FiscalPeriod, $filter) {
   };
 
   // fires the onSelectCallback bound to the component boundary
-  $ctrl.onSelect = function ($item) {
-    $ctrl.onSelectCallback({ period : $item });
-  };
-}
\ No newline at end of file
+  $ctrl.onSelect = period => $ctrl.onSelectCallback({ period });
```
Comment (mbayopanda) [inline client/src/js/components/bhReportPeriodSelect.js:52]: 👍

File: server/controllers/finance/reports/balance/index.js [hunk 2]
```diff
@@ -59,164 +56,95 @@ function document(req, res, next) {
...
+        IF(p.id < ?, SUM(pt.debit - pt.credit), 0) AS "before",
+        IF(p.id = ?, SUM(pt.debit - pt.credit), 0) AS "during",
+        IF(p.id <= ?, SUM(pt.debit - pt.credit), 0) AS "after"
+      FROM period_total AS pt
+      JOIN period AS p ON p.id = pt.period_id
+      WHERE pt.fiscal_year_id = ?
+      GROUP BY pt.account_id
+    )s ON a.id = s.account_id
+    ORDER BY a.number;
+  `;
```
Comment (mbayopanda) [inline server/controllers/finance/reports/balance/index.js:103]: 💃 good

**Review by mbayopanda — COMMENTED** [review]
The code LGTM and the report works well.

I noticed that the totals are not displayed on the bottom of the table.

balance report

**Comment by jniles** [comment]
@mbayopanda, changes have landed!

**Comment by jniles** [comment]
@mbayopanda, changes have landed!

**Comment by jniles** [comment]
Right, this has been over a week and it got a LGTM originally.

bors r+

**Pull request merged**